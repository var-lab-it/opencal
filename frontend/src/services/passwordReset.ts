import {EventTypes} from "../types/EventType";
import apiClient from "./api";

export async function requestPassword(email: string): Promise<null> {
    try {
        const response = await apiClient.get(`/event_types?&host.email=${email}`);
        return response.data;
    } catch (error) {
        console.error(error);
        throw error;
    }
}
