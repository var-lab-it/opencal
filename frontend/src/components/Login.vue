<template>
  <div class="login-form">
    <h2>Login</h2>
    <form @submit.prevent="handleLogin">
      <div class="form-group mb-3">
        <label for="email">Email:</label>
        <input
          id="email"
          v-model="email"
          type="email"
          class="form-control"
          required
          data-testid="email-input"
        />
      </div>
      <div class="form-group mb-3">
        <label for="password">Password:</label>
        <input
          id="password"
          v-model="password"
          type="password"
          class="form-control"
          required
          data-testid="password-input"
        />
      </div>
      <div class="d-grid">
        <button
          type="submit"
          class="btn btn-primary"
          data-testid="login-btn"
        >
          Login
        </button>
      </div>
    </form>
    <p
      v-if="error"
      class="text-danger mt-3"
    >
      {{ error }}
    </p>
  </div>
</template>

<script setup lang="ts">
import {ref} from "vue";
import apiClient from "../services/api";
import {redirectAfterLogin} from "../services/auth";

const email = ref('');
const password = ref('');
const error = ref('');

async function handleLogin() {
  error.value = '';
  try {
    const response = await apiClient.post('/auth', {
      email: email.value,
      password: password.value,
    });

    if (response.data && response.data.token) {
      const token = response.data.token;

      sessionStorage.setItem('jwtToken', token);
      redirectAfterLogin();
    } else {
      error.value = 'No valid token.';
    }
  } catch (err) {
    console.error("Login failed:", err);
    error.value = "Login failed. Please try again.";
  }
}
</script>

<style scoped>
.login-form {
  max-width: 400px;
  margin: 0 auto;
  padding: 20px;
  border: 1px solid #ddd;
  border-radius: 10px;
}
</style>
