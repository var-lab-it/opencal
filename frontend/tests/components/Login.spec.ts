import {describe, expect, it, vi} from "vitest";
import Login from "../../src/components/Login.vue";
import apiClient from "../../src/services/api";
import {render} from "vitest-browser-vue";
import {redirectAfterLogin} from "../../src/services/auth";

vi.mock("../../src/services/api", () => ({
    default: {
        post: vi.fn(),
    },
}));
vi.mock("../../src/services/auth", () => ({
    redirectAfterLogin: vi.fn(),
}));

describe("Login.vue", () => {
    it("renders correctly and displays all necessary elements", () => {
        const result = render(Login);

        expect(result).toMatchSnapshot();
    });

    it("displays an error message when login fails", async () => {
        const {getByTestId, getByText} = render(Login);

        (apiClient.post as vi.Mock)
            .mockRejectedValueOnce(new Error("Error"));

        const emailInput = getByTestId('email-input');
        const passwordInput = getByTestId('password-input');
        const loginBtn = getByTestId('login-btn');

        await emailInput.fill("test@example.com");
        await passwordInput.fill("12345678");
        await loginBtn.click();

        expect(getByText("Login failed. Please try again.")).toBeTruthy();
    });

    it("stores the token in sessionStorage and redirects on successful login", async () => {
        const { getByTestId } = render(Login);

        const token = "test-token";

        (apiClient.post as vi.Mock).mockResolvedValueOnce({
            data: { token: token },
        });

        (redirectAfterLogin as vi.Mock).mockResolvedValueOnce();

        const emailInput = getByTestId('email-input');
        const passwordInput = getByTestId('password-input');
        const loginBtn = getByTestId('login-btn');

        await emailInput.fill("test@example.com");
        await passwordInput.fill("12345678");
        await loginBtn.click();
    });

    it("displays an error message when the token is missing in the response", async () => {
        const { getByText, getByTestId } = render(Login);

        (apiClient.post as vi.Mock).mockResolvedValueOnce({
            data: {},
        });

        const emailInput = getByTestId('email-input');
        const passwordInput = getByTestId('password-input');
        const loginBtn = getByTestId('login-btn');

        await emailInput.fill("test@example.com");
        await passwordInput.fill("12345678");
        await loginBtn.click();

        expect(getByText("No valid token.")).toBeTruthy();
    });
});
