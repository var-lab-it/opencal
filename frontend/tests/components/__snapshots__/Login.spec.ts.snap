// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login.vue > renders correctly and displays all necessary elements 1`] = `
{
  "baseElement": <body>
    
  


    <div>
      <div
        class="login-form"
        data-v-47199182=""
      >
        <h2
          data-v-47199182=""
        >
          Login
        </h2>
        <form
          data-v-47199182=""
        >
          <div
            class="form-group mb-3"
            data-v-47199182=""
          >
            <label
              data-v-47199182=""
              for="email"
            >
              Email:
            </label>
            <input
              class="form-control"
              data-testid="email-input"
              data-v-47199182=""
              id="email"
              required=""
              type="email"
            />
          </div>
          <div
            class="form-group mb-3"
            data-v-47199182=""
          >
            <label
              data-v-47199182=""
              for="password"
            >
              Password:
            </label>
            <input
              class="form-control"
              data-testid="password-input"
              data-v-47199182=""
              id="password"
              required=""
              type="password"
            />
          </div>
          <div
            class="d-grid"
            data-v-47199182=""
          >
            <button
              class="btn btn-primary"
              data-testid="login-btn"
              data-v-47199182=""
              type="submit"
            >
              <!--v-if-->
               Login 
            </button>
          </div>
        </form>
        <!--v-if-->
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-form"
      data-v-47199182=""
    >
      <h2
        data-v-47199182=""
      >
        Login
      </h2>
      <form
        data-v-47199182=""
      >
        <div
          class="form-group mb-3"
          data-v-47199182=""
        >
          <label
            data-v-47199182=""
            for="email"
          >
            Email:
          </label>
          <input
            class="form-control"
            data-testid="email-input"
            data-v-47199182=""
            id="email"
            required=""
            type="email"
          />
        </div>
        <div
          class="form-group mb-3"
          data-v-47199182=""
        >
          <label
            data-v-47199182=""
            for="password"
          >
            Password:
          </label>
          <input
            class="form-control"
            data-testid="password-input"
            data-v-47199182=""
            id="password"
            required=""
            type="password"
          />
        </div>
        <div
          class="d-grid"
          data-v-47199182=""
        >
          <button
            class="btn btn-primary"
            data-testid="login-btn"
            data-v-47199182=""
            type="submit"
          >
            <!--v-if-->
             Login 
          </button>
        </div>
      </form>
      <!--v-if-->
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
